### PR TITLE
Add SnoopCompile, part 1

### LINKED ISSUE #4: Use Cthulhu v2.0 and new SnoopCompile version
Now Cthulhu v2.0 is released, which is compatible with Julia 1.7 and higher.
We still need to update SnoopCompile too, because the current latest version of it is compatible with Cthulhu v1.5.

### PR DESCRIPTION
**Pull request by timholy**
This does not include the inference triggers functionality, it's mostly focused on the GUIs.

Part 2 will run after Cthulhu and focus on fixing inference problems.

Fixes #4

### CODE REVIEW

File: SnoopCompileBasics.ipynb [hunk 1]
```diff
@@ -0,0 +1,717 @@
...
+   },
+   "source": [
+    "```julia\n",
+    "sum2(list) = [sum(list[1]), sum(list[2])]\n",
+    "\n",
+    "data = Any[[1,2,3], Any[1,2,3]]\n",
+    "\n",
+    "sum2(data)\n",
+    "\n",
+    "# sum(Any[1,2,3])\n",
```
Comment (aviatesk) [inline SnoopCompileBasics.ipynb:104]:
```suggestion
    "# sum2(Any[[1,2,3], []])\n",
```
maybe ?

**Review by aviatesk — COMMENTED** [review]
I also found `flamegraph(tinf)` results in error (both on v1.6 and v1.7)
```julia
julia> using Flux, SnoopCompile

julia> # From Flux's introductory documentation
       actual(x) = 4x + 2
actual (generic function with 1 method)

julia> loss(predict, x, y) = Flux.Losses.mse(predict(x), y)
loss (generic function with 1 method)

julia> x_train, x_test = hcat(0:5...), hcat(6:10...)
([0 1 … 4 5], [6 7 … 9 10])

julia> y_train, y_test = actual.(x_train), actual.(x_test)
([2 6 … 18 22], [26 30 … 38 42])

julia> tinf = tinfflux = @snoopi_deep begin
           predict = Dense(1, 1)
           parameters = params(predict)
           Flux.train!((x, y) -> loss(predict, x, y), parameters, [(x_train, y_train)], Descent())
       end
InferenceTimingNode: 11.496437/28.881830 on InferenceFrameInfo for Core.Compiler.Timings.ROOT() with 409 direct children

julia> flamegraph(tinf)
ERROR: MethodError: no method matching known_type(::Module, ::TypeVar)
Closest candidates are:
  known_type(::Module, ::Type) at /Users/aviatesk/.julia/packages/SnoopCompile/DrloO/src/parcel_snoopi.jl:197
Stacktrace:
 [1] known_type(mod::Module, T::Type)
   @ SnoopCompile ~/.julia/packages/SnoopCompile/DrloO/src/parcel_snoopi.jl:219
 [2] isprecompilable(mi::Core.MethodInstance; excluded_modules::Set{Module})
   @ SnoopCompile ~/.julia/packages/SnoopCompile/DrloO/src/parcel_snoopi_deep.jl:264
 [3] _flamegraph_frame(io::IOBuffer, node::SnoopCompileCore.InferenceTimingNode, start_secs::Float64, check_precompilable::Bool, excluded_modules::Set{Module}, mode::Nothing; toplevel::Bool)
   @ SnoopCompile ~/.julia/packages/SnoopCompile/DrloO/src/parcel_snoopi_deep.jl:1756
 [4] _build_flamegraph!(root::LeftChildRightSiblingTrees.Node{FlameGraphs.NodeData}, io::IOBuffer, node::SnoopCompileCore.InferenceTimingNode, start_secs::Float64, tmin::Float64, check_precompilable::Bool, excluded_modules::Set{Module}, mode::Nothing)
   @ SnoopCompile ~/.julia/packages/SnoopCompile/DrloO/src/parcel_snoopi_deep.jl:1711
 [5] _build_flamegraph!(root::LeftChildRightSiblingTrees.Node{FlameGraphs.NodeData}, io::IOBuffer, node::SnoopCompileCore.InferenceTimingNode, start_secs::Float64, tmin::Float64, check_precompilable::Bool, excluded_modules::Set{Module}, mode::Nothing) (repeats 7 times)
   @ SnoopCompile ~/.julia/packages/SnoopCompile/DrloO/src/parcel_snoopi_deep.jl:1713
 [6] flamegraph(tinf::SnoopCompileCore.InferenceTimingNode; tmin::Float64, excluded_modules::Set{Module}, mode::Nothing)
   @ SnoopCompile ~/.julia/packages/SnoopCompile/DrloO/src/parcel_snoopi_deep.jl:1706
 [7] flamegraph(tinf::SnoopCompileCore.InferenceTimingNode)
   @ SnoopCompile ~/.julia/packages/SnoopCompile/DrloO/src/parcel_snoopi_deep.jl:1697
 [8] top-level scope
   @ REPL[9]:1
```

File: SnoopCompileBasics.ipynb (hunk 1 above)
Comment (timholy) [reply SnoopCompileBasics.ipynb:104]: It's there to show the line that JET does pick up on.

**Comment by timholy** [comment]
Error fixed locally, PR coming soon (I'm testing Cthulhu 2.0 integration first...).

**Comment by timholy** [comment]
https://github.com/timholy/SnoopCompile.jl/pull/250

**Review by aviatesk — COMMENTED** [review]
Now can we bump the versions in our project as well ?

**Pull request merged**